src/ssm-mem.c

```diff
@@ -54,30 +54,56 @@ struct ssm_object *ssm_new(uint8_t val_count, uint8_t tag) {
 void ssm_dup(struct ssm_mm *mm) { ++mm->ref_count; }
 
 void ssm_drop(struct ssm_mm *mm) {
-  if (--mm->ref_count == 0)
+  if (--mm->ref_count == 0) {
+    if (!ssm_mm_is_builtin(mm)) {
+      struct ssm_object *obj = container_of(mm, struct ssm_object, mm);
+      for (int i = 0; i < mm->val_count; i++) {
+        if (!ssm_mm_is_builtin(obj->payload[i].heap_ptr)) {
+          ssm_drop(obj->payload[i].heap_ptr);
+        }
+      }
+    }
     ssm_mem_free(mm, ssm_mm_is_builtin(mm) ? SSM_BUILTIN_SIZE(mm->tag)
                                            : SSM_OBJ_SIZE(mm->val_count));
+  }
 }
 
 struct ssm_mm *ssm_reuse(struct ssm_mm *mm) {
-  SSM_ASSERT(0); // TODO: not implemented
-  return NULL;
+  if (--mm->ref_count == 0) {
+    if (!ssm_mm_is_builtin(mm)) {
+      struct ssm_object *obj = container_of(mm, struct ssm_object, mm);
+      for (int i = 0; i < mm->val_count; i++) {
+        if (!ssm_mm_is_builtin(obj->payload[i].heap_ptr)) {
+          ssm_drop(obj->payload[i].heap_ptr);
+        }
+      }
+    }
+    return mm;
+  } else {
+    if (ssm_mm_is_builtin(mm)) {
+      return ssm_new_builtin(mm->tag);
+    } else {
+      return ssm_new(mm->val_count, mm->tag);
+    }
+  }
 }
 
 allocation_dispatcher_t *ad_initialize(size_t block_sizes[],
                                        size_t num_blocks[],
                                        size_t num_allocators,
                                        void *memory_pool) {
-  void *memoryHead = memory_pool;
-  allocation_dispatcher_t *dispatcher = memoryHead;
-  memoryHead = ((char *)memoryHead) + sizeof(allocation_dispatcher_t);
+  // memory head is keeping track of next available memory
```

Comment:
was this meant to be part of this PR?

Reply:
oh, you're merging into your own branch

Reply:
you're right, this was supposed to be part of the other branch

Reply:
https://github.com/ssm-lang/ssm-runtime/pull/5/files#diff-bf26217fc289609e08d522bde9ef62c62d3e6c7cf71fbeb74d220b7ea1677aceR71 it's there too; weird github bug probably